tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1305 @@
+    /// [CreateFile]: https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew
+    pub fn write(&mut self, allowed: bool) -> &mut Self {
+        bool_flag!(self.desired_access, allowed, winnt::GENERIC_WRITE);
+        self
+    }
+
+    /// Open the named pipe identified by `addr`.
+    ///
+    /// This constructs the handle using [CreateFile].
+    ///
+    /// [CreateFile]: https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea
+    ///
+    /// # Errors
+    ///
+    /// This errors if called outside of a [Tokio Runtime] which doesn't have
+    /// [I/O enabled] or if any OS-specific I/O errors occur.
+    ///
+    /// There are a few errors you should be aware of that you need to take into
+    /// account when creating a named pipe on the client side:
+    ///
+    /// * [std::io::ErrorKind::NotFound] - This indicates that the named pipe
+    ///   does not exist. Presumably the server is not up.
+    /// * [ERROR_PIPE_BUSY] - which needs to be tested for through a constant in
+    ///   [winapi]. This error is raised when the named pipe has been created,
+    ///   but the server is not currently waiting for a connection.
+    ///
+    /// [ERROR_PIPE_BUSY]: crate::winapi::shared::winerror::ERROR_PIPE_BUSY
+    /// [I/O enabled]: crate::runtime::Builder::enable_io
+    /// [Tokio Runtime]: crate::runtime::Runtime
+    /// [winapi]: crate::winapi
+    ///
+    /// A connect loop that waits until a socket becomes available looks like
+    /// this:
+    ///
+    /// ```no_run
+    /// use std::time::Duration;
+    /// use tokio::net::windows::NamedPipeClientOptions;
+    /// use tokio::time;
+    /// use winapi::shared::winerror;
+    ///
+    /// const PIPE_NAME: &str = r"\\.\pipe\mynamedpipe";
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let client = loop {
+    ///     match NamedPipeClientOptions::new().create(PIPE_NAME) {
+    ///         Ok(client) => break client,
+    ///         Err(e) if e.raw_os_error() == Some(winerror::ERROR_PIPE_BUSY as i32) => (),
+    ///         Err(e) => return Err(e),
+    ///     }
+    ///
+    ///     time::sleep(Duration::from_millis(50)).await;
+    /// };
+    ///
+    /// // use the connected client.
+    /// # Ok(()) }
+    /// ```
+    pub fn create(&self, addr: impl AsRef<OsStr>) -> io::Result<NamedPipe> {
+        // Safety: We're calling create_with_security_attributes w/ a null
+        // pointer which disables it.
+        unsafe { self.create_with_security_attributes(addr, ptr::null_mut()) }
+    }
+
+    /// Open the named pipe identified by `addr`.
+    ///
+    /// This is the same as [create][NamedPipeClientOptions::create] except that
+    /// it supports providing security attributes.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `attrs` points to an initialized instance
+    /// of a [SECURITY_ATTRIBUTES] structure.
```

Comment:
Null pointers are also allowed.

Reply:
They are, but we'd also like them to use `create` instead if they don't need it :)

Reply:
added a note